File: internal/ast/mime.go
package ast

import (
	"fmt"
	"strings"
)

type MIME uint8

const (
	MIMEUnknown MIME = iota

	// Image
	MIMEJpeg
	MIMEPng
	MIMEGif
	MIMEWebp
	MIMESvg
	MIMEAvif
	MIMEBmp
	MIMEIco
	MIMETiff
	MIMEHeic
	MIMEHeif

	// Text
	MIMETextPlain
	MIMEHtml
	MIMECss
	MIMEJavaScript
	MIMEJson
	MIMECsv
	MIMEMarkdown

	// Application
	MIMEPdf
	MIMEZip
	MIMEGzip
	MIMETar
	MIMEXlsx
	MIMEDocx
	MIMEOctetStream

	// Video
	MIMEMp4
	MIMEWebm
	MIMEMov

	// Audio
	MIMEMp3
	MIMEWav
	MIMEFlac

	// Font
	MIMEWoff2
	MIMETtf
)

func (m MIME) String() string {
	switch m {
	// Image
	case MIMEJpeg:
		return "image/jpeg"
	case MIMEPng:
		return "image/png"
	case MIMEGif:
		return "image/gif"
	case MIMEWebp:
		return "image/webp"
	case MIMESvg:
		return "image/svg+xml"
	case MIMEAvif:
		return "image/avif"
	case MIMEBmp:
		return "image/bmp"
	case MIMEIco:
		return "image/x-icon"
	case MIMETiff:
		return "image/tiff"
	case MIMEHeic:
		return "image/heic"
	case MIMEHeif:
		return "image/heif"

	// Text
	case MIMETextPlain:
		return "text/plain"
	case MIMEHtml:
		return "text/html"
	case MIMECss:
		return "text/css"
	case MIMEJavaScript:
		return "text/javascript"
	case MIMEJson:
		return "application/json"
	case MIMECsv:
		return "text/csv"
	case MIMEMarkdown:
		return "text/markdown"

	// Application
	case MIMEPdf:
		return "application/pdf"
	case MIMEZip:
		return "application/zip"
	case MIMEGzip:
		return "application/gzip"
	case MIMETar:
		return "application/x-tar"
	case MIMEXlsx:
		return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
	case MIMEDocx:
		return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
	case MIMEOctetStream:
		return "application/octet-stream"

	// Video
	case MIMEMp4:
		return "video/mp4"
	case MIMEWebm:
		return "video/webm"
	case MIMEMov:
		return "video/mov"

	// Audio
	case MIMEMp3:
		return "audio/mpeg"
	case MIMEWav:
		return "audio/wav"
	case MIMEFlac:
		return "audio/flac"

	// Font
	case MIMEWoff2:
		return "font/woff2"
	case MIMETtf:
		return "font/ttf"

	default:
		return "unknown"
	}
}

var strToMIME = map[string]MIME{
	"image/jpeg":             MIMEJpeg,
	"image/jpg":              MIMEJpeg,
	"image/png":              MIMEPng,
	"image/gif":              MIMEGif,
	"image/webp":             MIMEWebp,
	"image/svg+xml":          MIMESvg,
	"image/avif":             MIMEAvif,
	"image/bmp":              MIMEBmp,
	"image/x-icon":           MIMEIco,
	"image/tiff":             MIMETiff,
	"image/heic":             MIMEHeic,
	"image/heif":             MIMEHeif,
	"text/plain":             MIMETextPlain,
	"text/html":              MIMEHtml,
	"text/css":               MIMECss,
	"text/javascript":        MIMEJavaScript,
	"application/javascript": MIMEJavaScript,
	"application/json":       MIMEJson,
	"text/csv":               MIMECsv,
	"text/markdown":          MIMEMarkdown,
	"application/pdf":        MIMEPdf,
	"application/zip":        MIMEZip,
	"application/gzip":       MIMEGzip,
	"application/x-tar":      MIMETar,
	"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet":       MIMEXlsx,
	"application/vnd.openxmlformats-officedocument.wordprocessingml.document": MIMEDocx,
	"application/octet-stream": MIMEOctetStream,
	"video/mp4":                MIMEMp4,
	"video/webm":               MIMEWebm,
	"video/mov":                MIMEMov,
	"audio/mpeg":               MIMEMp3,
	"audio/wav":                MIMEWav,
	"audio/flac":               MIMEFlac,
	"font/woff2":               MIMEWoff2,
	"font/ttf":                 MIMETtf,
}

func ParseMIME(s string) (MIME, error) {
	s = strings.ToLower(strings.TrimSpace(s))
	if m, ok := strToMIME[s]; ok {
		return m, nil
	}
	return MIMEUnknown, fmt.Errorf("Invalid MIME string: %s", s)
}

// IsImage
func (m MIME) IsImage() bool {
	switch m {
	case MIMEJpeg, MIMEPng, MIMEGif, MIMEWebp, MIMESvg, MIMEAvif, MIMEBmp, MIMEIco, MIMETiff, MIMEHeic, MIMEHeif:
		return true
	default:
		return false
	}
}

// Ext
func (m MIME) Ext() string {
	switch m {
	case MIMEJpeg:
		return "jpeg"
	case MIMEPng:
		return "png"
	case MIMEGif:
		return "gif"
	case MIMEWebp:
		return "webp"
	case MIMESvg:
		return "svg"
	case MIMEAvif:
		return "avif"
	case MIMEBmp:
		return "bmp"
	case MIMEIco:
		return "ico"
	case MIMETiff:
		return "tiff"
	case MIMEHeic:
		return "heic"
	case MIMEHeif:
		return "heif"
	case MIMETextPlain:
		return "txt"
	case MIMEHtml:
		return "html"
	case MIMECss:
		return "css"
	case MIMEJavaScript:
		return "js"
	case MIMEJson:
		return "json"
	case MIMECsv:
		return "csv"
	case MIMEMarkdown:
		return "md"
	case MIMEPdf:
		return "pdf"
	case MIMEZip:
		return "zip"
	case MIMEGzip:
		return "gz"
	case MIMETar:
		return "tar"
	case MIMEXlsx:
		return "xlsx"
	case MIMEDocx:
		return "docx"
	case MIMEMp4:
		return "mp4"
	case MIMEWebm:
		return "webm"
	case MIMEMov:
		return "mov"
	case MIMEMp3:
		return "mp3"
	case MIMEWav:
		return "wav"
	case MIMEFlac:
		return "flac"
	case MIMEWoff2:
		return "woff2"
	case MIMETtf:
		return "ttf"
	default:
		return ""
	}
}

File: mm-py/metamessage/ir/mime.py
from typing import Optional
from enum import IntEnum


class MIME(IntEnum):
    Unknown = 0

    Jpeg = 1
    Png = 2
    Gif = 3
    Webp = 4
    Svg = 5
    Avif = 6
    Bmp = 7
    Ico = 8
    Tiff = 9
    Heic = 10
    Heif = 11

    TextPlain = 12
    Html = 13
    Css = 14
    JavaScript = 15
    Json = 16
    Csv = 17
    Markdown = 18

    Pdf = 19
    Zip = 20
    Gzip = 21
    Tar = 22
    Xlsx = 23
    Docx = 24
    OctetStream = 25

    Mp4 = 26
    Webm = 27
    Mov = 28

    Mp3 = 29
    Wav = 30
    Flac = 31

    Woff2 = 32
    Ttf = 33

    def __str__(self) -> str:
        mapping = {
            MIME.Jpeg: "image/jpeg",
            MIME.Png: "image/png",
            MIME.Gif: "image/gif",
            MIME.Webp: "image/webp",
            MIME.Svg: "image/svg+xml",
            MIME.Avif: "image/avif",
            MIME.Bmp: "image/bmp",
            MIME.Ico: "image/x-icon",
            MIME.Tiff: "image/tiff",
            MIME.Heic: "image/heic",
            MIME.Heif: "image/heif",
            MIME.TextPlain: "text/plain",
            MIME.Html: "text/html",
            MIME.Css: "text/css",
            MIME.JavaScript: "text/javascript",
            MIME.Json: "application/json",
            MIME.Csv: "text/csv",
            MIME.Markdown: "text/markdown",
            MIME.Pdf: "application/pdf",
            MIME.Zip: "application/zip",
            MIME.Gzip: "application/gzip",
            MIME.Tar: "application/x-tar",
            MIME.Xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            MIME.Docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            MIME.OctetStream: "application/octet-stream",
            MIME.Mp4: "video/mp4",
            MIME.Webm: "video/webm",
            MIME.Mov: "video/mov",
            MIME.Mp3: "audio/mpeg",
            MIME.Wav: "audio/wav",
            MIME.Flac: "audio/flac",
            MIME.Woff2: "font/woff2",
            MIME.Ttf: "font/ttf",
        }
        return mapping.get(self, "unknown")

    def is_image(self) -> bool:
        return self in (
            MIME.Jpeg, MIME.Png, MIME.Gif, MIME.Webp, MIME.Svg, MIME.Avif,
            MIME.Bmp, MIME.Ico, MIME.Tiff, MIME.Heic, MIME.Heif,
        )

    def ext(self) -> str:
        mapping = {
            MIME.Jpeg: "jpeg",
            MIME.Png: "png",
            MIME.Gif: "gif",
            MIME.Webp: "webp",
            MIME.Svg: "svg",
            MIME.Avif: "avif",
            MIME.Bmp: "bmp",
            MIME.Ico: "ico",
            MIME.Tiff: "tiff",
            MIME.Heic: "heic",
            MIME.Heif: "heif",
            MIME.TextPlain: "txt",
            MIME.Html: "html",
            MIME.Css: "css",
            MIME.JavaScript: "js",
            MIME.Json: "json",
            MIME.Csv: "csv",
            MIME.Markdown: "md",
            MIME.Pdf: "pdf",
            MIME.Zip: "zip",
            MIME.Gzip: "gz",
            MIME.Tar: "tar",
            MIME.Xlsx: "xlsx",
            MIME.Docx: "docx",
            MIME.Mp4: "mp4",
            MIME.Webm: "webm",
            MIME.Mov: "mov",
            MIME.Mp3: "mp3",
            MIME.Wav: "wav",
            MIME.Flac: "flac",
            MIME.Woff2: "woff2",
            MIME.Ttf: "ttf",
        }
        return mapping.get(self, "")


_str_to_mime = {
    "image/jpeg": MIME.Jpeg,
    "image/jpg": MIME.Jpeg,
    "image/png": MIME.Png,
    "image/gif": MIME.Gif,
    "image/webp": MIME.Webp,
    "image/svg+xml": MIME.Svg,
    "image/avif": MIME.Avif,
    "image/bmp": MIME.Bmp,
    "image/x-icon": MIME.Ico,
    "image/tiff": MIME.Tiff,
    "image/heic": MIME.Heic,
    "image/heif": MIME.Heif,
    "text/plain": MIME.TextPlain,
    "text/html": MIME.Html,
    "text/css": MIME.Css,
    "text/javascript": MIME.JavaScript,
    "application/javascript": MIME.JavaScript,
    "application/json": MIME.Json,
    "text/csv": MIME.Csv,
    "text/markdown": MIME.Markdown,
    "application/pdf": MIME.Pdf,
    "application/zip": MIME.Zip,
    "application/gzip": MIME.Gzip,
    "application/x-tar": MIME.Tar,
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": MIME.Xlsx,
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document": MIME.Docx,
    "application/octet-stream": MIME.OctetStream,
    "video/mp4": MIME.Mp4,
    "video/webm": MIME.Webm,
    "video/mov": MIME.Mov,
    "audio/mpeg": MIME.Mp3,
    "audio/wav": MIME.Wav,
    "audio/flac": MIME.Flac,
    "font/woff2": MIME.Woff2,
    "font/ttf": MIME.Ttf,
}


def ParseMIME(s: str) -> int:
    s = s.lower().strip()
    m = _str_to_mime.get(s, MIME.Unknown)
    return int(m)